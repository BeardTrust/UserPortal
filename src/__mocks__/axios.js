export default {
    post: jest.fn().mockResolvedValue({
        response: {
            data: {},
            headers: {}
        }
    }),
    put: jest.fn().mockResolvedValue({
        response: {
            data: {}
        }
    }),
    get: jest.fn().mockResolvedValue({
        data: {
            user: {
                userId: 'abc-123-xyz-789',
                email: 'a@b.com',
                username: '*',
                password: '*',
                firstName: '*',
                lastName: '*',
                phone: '3',
                dateOfBirth: '1234-01-01'
            }
        }
    })
}