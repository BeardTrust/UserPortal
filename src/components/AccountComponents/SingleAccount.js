import { Button, Table, Modal } from "react-bootstrap"
import { useContext, useRef, useState } from "react"
import AuthContext from "../../store/auth-context"
import axios from "axios"
import Deactivator from "./AccountDeactivation/AccountDeactivator";
import { render } from "@testing-library/react";
import { useHistory } from "react-router";
import { useEffect } from "react";

const SingleAccount = ({ accounts }) => {

    var [account, setAccount] = useState({});
    var [amount, setAmount] = useState({});
    const [show, setShow] = useState(false);
    const history = useHistory();
    useEffect(() => {
        setAccount(accounts)
    }, [accounts])

    const handleClose = () => setShow(false);
    const handleShow = () => setShow(true);
    const authContext = useContext(AuthContext);
    const token = authContext.token;
    const withAmt = useRef();
    const depAmt = useRef();
    const TransferEntity = { amount };

    function deactivateHandler(account, history) {
        render(
            <section>
                <Deactivator account={account} history={history} />
            </section>
        )
    }

    if (accounts === null) {
        return null
    } else {
        return (
            <div>
                <Table striped bordered hover>
                    <thead>
                        <tr>
                            <th>Nickname</th>
                            <th>Balance</th>
                            <th>Interest</th>
                            <th>Date Created</th>
                            <th>Type</th>
                            <th>Withdraw</th>
                            <th>Deposit</th>
                            <th>Deactivate</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>{account ? account.nickname : null}</td>
                            <td>${account.balance ? account?.balance['dollars'] + '.' + account?.balance['cents'] : null}</td>
                            <td>{account ? account.interest : null}%</td>
                            <td>{account ? account.create_date : null}</td>
                            <td>{account ? account.type : null}</td>
                            <td>$<input
                                type="text"
                                id="withdrawInput"
                                ref={withAmt} />
                                <Button
                                    variant="success"
                                    type={'submit'}
                                    onClick={submitWithdraw}
                                    title='withdrawButton'
                                >Withdraw</Button></td>
                            <td>$<input
                                type="int"
                                id="depositInput"
                                ref={depAmt} />
                                <Button
                                    variant="success"
                                    type={'submit'}
                                    onClick={submitDeposit}
                                    title='depositButton'
                                >Deposit</Button></td>
                            <td><Button
                                onClick={handleShow}
                                variant="danger"
                                type={'submit'}
                                title='deactivate'>
                                Deactivate Account</Button>

                                <Modal show={show} onHide={handleClose}>
                                    <Modal.Header closeButton>
                                        <Modal.Title>Are You Sure?</Modal.Title>
                                    </Modal.Header>
                                    <Modal.Body>Once deactivated, you will no longer be able to use this account: {account ? account.nickname : null}</Modal.Body>
                                    <Modal.Footer>
                                        <Button variant="secondary" onClick={handleClose}>
                                            Cancel
                                        </Button>
                                        <Button variant="primary" onClick={() => deactivateHandler(account, history)}>
                                            Deactivate Account
                                        </Button>
                                    </Modal.Footer>
                                </Modal></td>
                        </tr>
                    </tbody>
                </Table>
            </div>
        )
    }

    function submitWithdraw(event) {
        if (!event === null) {
            event.preventDefault();
        }
        console.log('withdrw value:', withAmt.current.value);
        let amount = parseInt(withAmt.current.value, 10);
        if (amount === parseInt(amount, 10)) {
            if (amount > 0) {
                amount *= -1
            }
            changeMoney(amount)
        } else {
            console.log('amount not an integer')
        }

    }

    function submitDeposit(event) {
        if (!event === null) {
            event.preventDefault();
        }
        console.log('dpst value:', depAmt.current.value);
        let amount = parseInt(depAmt.current.value, 10);
        if (amount === parseInt(amount, 10)) {
            if (amount < 0) {
                amount *= -1
            }
            changeMoney(amount);
        } else {
            console.log('amount not an integer')
        }

    }

    async function changeMoney(amount) {
        TransferEntity.amount = amount
        const url = 'http://localhost:9001/accounts/' + account.accountId
        const headers = {
            'Authorization': token,
            'Content-Type': 'application/json'
        };

        try {
            const response = await axios.put(url, TransferEntity);
            console.log(response.data)
            setAccount(response.data)
            window.location.reload();
        } catch (e) {
            console.log(e)
        }
    }
}

export default SingleAccount