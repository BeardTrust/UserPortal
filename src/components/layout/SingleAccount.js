import {Button, Table} from "react-bootstrap"
import {useContext, useRef, useState} from "react"
import AuthContext from "../../store/auth-context"
import axios from "axios"

const SingleAccount = ({accounts}) => {

    var [account, setAccount] = useState({});
    var [amount, setAmount] = useState({});
    const authContext = useContext(AuthContext);
    const token = authContext.token;
    const withAmt = useRef();
    const depAmt = useRef();
    let dispActs = []
    const TransferEntity = { amount: amount };

    if (!Array.prototype.slice.call(accounts).length === 0) {
        console.log('account checks as empty, dispActs: ', dispActs)
        return null
    } else {
        console.log('accounts has more than 0, dispActs: ', dispActs)
        dispActs = [].slice.call(accounts)
        console.log('after dispActs is set: ', dispActs)

        return (
            <div>
                <Table striped bordered hover>
                    <thead>
                    <tr>
                        <th>Nickname</th>
                        <th>Balance</th>
                        <th>Interest</th>
                        <th>Date Created</th>
                        <th>Type</th>
                        <th>Withdraw</th>
                        <th>Deposit</th>
                    </tr>
                    </thead>
                    <tbody>
                    {(dispActs ?? []).map((account, index) => (
                        <tr data-index={index}>
                            <td>{account.nickname}</td>
                            <td>${account.balance}</td>
                            <td>{account.interest}%</td>
                            <td>{account.create_date}</td>
                            <td>{account.type}</td>
                            <td>$<input
                                type="text"
                                id="withdraw"
                                ref={withAmt}/>
                                <Button
                                    variant="success"
                                    type={'submit'}
                                    onClick={submitWithdraw}
                                    id='Review'
                                >Withdraw</Button></td>
                            <td>$<input
                                type="int"
                                id="deposit"
                                ref={depAmt}/>
                                <Button
                                    variant="success"
                                    type={'submit'}
                                    onClick={submitDeposit}
                                    id='Review'
                                >Deposit</Button></td>
                        </tr>
                    ))}
                    </tbody>
                </Table>
            </div>
        )
    }

    function submitWithdraw(event) {
        event.preventDefault();
        console.log('withdraw detected')
        let amount = parseInt(withAmt.current.value, 10)
        if (amount === parseInt(amount, 10)) {
            if (amount > 0) {
                amount *= -1
            }
            setAmount(parseInt(amount))
            changeMoney()
        } else {
            console.log('amount not an integer')
        }

    }

    function submitDeposit(event) {
        event.preventDefault();
        console.log('deposit detected')
        let amount = parseInt(depAmt.current.value, 10)
        console.log(amount)
        if (amount === parseInt(amount, 10)) {
            if (amount < 0) {
                amount *= -1
            }
            setAmount(parseInt(amount))
            changeMoney();
        } else {
            console.log('amount not an integer')
        }

    }

    async function changeMoney(event) {
        const url = 'http://localhost:9001/accounts/' + accounts[0].accountId
        if (!event === null) {
            event.preventDefault();
        }

        const headers = {
            'Authorization': token,
            'Content-Type': 'application/json'
        };

        try {
            const response = await axios.put(url, TransferEntity, headers);
            console.log(response)
        } catch (e) {
            console.log(e)
        }
    }
}

export default SingleAccount