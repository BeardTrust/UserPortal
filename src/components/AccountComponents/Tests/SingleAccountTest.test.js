import userEvent from "@testing-library/user-event";
import axios from "axios";
import TestRenderer from 'react-test-renderer'
import SingleAccount from '../SingleAccountDisplay'
import { render, screen } from '@testing-library/react';

jest.mock('axios')
describe('View Single Account ability', () => {

    const data = {
        data: {
            account: {
                userId: 'cf8bd72a-4b4c-42f1-8dc3-06cc2dc2cb8a',
                accountId: 'cf8bd72a-4b4c-42f1-8dc3-06cc2dc2cb8a',
                balance: '1000',
                interest: '1',
                nickname: 'name',
                type: 'Recovery',
                activeStatus: 'true',
                createDate: '2001-01-01'
            }
        }
    };

    it('Withdraws money from an account', async () => {
        // const account = axios.get.mockImplementationOnce(() => Promise.resolve(data));
        // const singleAccount = render(<SingleAccount accounts={account} url={'/test'} />);
        // const withdrawButton = screen.getByTitle('withdrawButton');
        // userEvent.click(withdrawButton);
        // expect(axios.get).toBeCalledTimes(1);
    });

    it('Deposits money into an account', async () => {
        // const account = axios.get.mockImplementationOnce(() => Promise.resolve(data));
        // const singleAccount = render(<SingleAccount accounts={account} url={'/test'} />);
        // const depositButton = screen.getByTitle('depositButton');
        // userEvent.click(depositButton);
        // expect(axios.get).toBeCalledTimes(1);
    });

    it('Deactivates an account', async () => {
        const account = axios.get.mockImplementationOnce(() => Promise.resolve(data));
        // const singleAccount = render(<SingleAccount accounts={account} url={'/test'} />);
        // const deactivateButton = screen.getByText('Deactivate Account');
        // userEvent.click(deactivateButton);
        // expect(axios.get).toBeCalledTimes(1);
    });

    it('Expects the snapshot to match the display', async () => {
        // const tr = TestRenderer.create(<SingleAccount/>);
        // expect(tr.toJSON()).toMatchSnapshot();
    });
})