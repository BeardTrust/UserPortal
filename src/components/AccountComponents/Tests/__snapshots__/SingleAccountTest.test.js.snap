// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Single Account ability Expects the snapshot to match the display 1`] = `
<div>
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Nickname
        </th>
        <th>
          Balance
        </th>
        <th>
          Interest
        </th>
        <th>
          Date Created
        </th>
        <th>
          Type
        </th>
        <th>
          Withdraw
        </th>
        <th>
          Deposit
        </th>
        <th>
          Deactivate
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td />
        <td>
          $
        </td>
        <td>
          %
        </td>
        <td />
        <td />
        <td>
          $
          <input
            id="withdrawInput"
            type="text"
          />
          <button
            className="btn btn-success"
            disabled={false}
            onClick={[Function]}
            title="withdrawButton"
            type="submit"
          >
            Withdraw
          </button>
        </td>
        <td>
          $
          <input
            id="depositInput"
            type="int"
          />
          <button
            className="btn btn-success"
            disabled={false}
            onClick={[Function]}
            title="depositButton"
            type="submit"
          >
            Deposit
          </button>
        </td>
        <td>
          <button
            className="btn btn-danger"
            disabled={false}
            onClick={[Function]}
            title="deactivate"
            type="submit"
          >
            Deactivate Account
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
